monkey/monkey_island/cc/repository/mongo_machine_repository.py

```diff
@@ -0,0 +1,102 @@
+from ipaddress import IPv4Address
+from threading import Lock
+from typing import Any, MutableMapping, Sequence
+
+from pymongo import MongoClient
+
+from common.types import HardwareID
+from monkey_island.cc.models import Machine, MachineID
+
+from . import IMachineRepository, RemovalError, RetrievalError, StorageError, UnknownRecordError
+from .consts import MONGO_OBJECT_ID_KEY
+
+
+class MongoMachineRepository(IMachineRepository):
+    """A repository used to store and retrieve Machines in MongoDB"""
+
+    def __init__(self, mongo_client: MongoClient):
+        self._machines_collection = mongo_client.monkey_island.machines
+        self._id_lock = Lock()
+        self._next_id = self._get_biggest_id()
+
+    def _get_biggest_id(self) -> MachineID:
+        try:
+            return self._machines_collection.find().sort("id", -1).limit(1)[0]["id"]
+        except IndexError:
+            return 0
+
+    def create_machine(self) -> Machine:
```

Comment:
Machine object probably doesn't need an ID. ID is relevant to a db entry. This effectively is a Machine constructor in a repository. I would consider just saving machine object as a document. Upon saving we check if there's an ID, if not mongo will create one. Then we either modify the object by reference of return a modified one

Reply:
We can cast to a document for complex repositories, but not sure it's needed for this one: https://pydantic-docs.helpmanual.io/usage/models/#orm-mode-aka-arbitrary-class-instances

Reply:
See https://github.com/guardicore/monkey/pull/2228#discussion_r959646019